duti.1
.TH duti "1" "_DUTI_BUILD_DATE" "Andrew Mortensen" "User Commands"
.SH NAME
.B duti
\- set default document and URL handlers
.SH SYNOPSIS
.B duti
[
.BI \-hVv
] [
.BI \-d\  uti
] [
.BI \-l\  uti
] [
.I settings_path
]
.sp
.B duti
.BI \-s
.I bundle_id
{
.I uti
|
.I url_scheme
|
.I extension
|
.I MIME_type
}
[
.I role
]
.sp
.B duti
.BI \-x\  extension
.sp
.SH DESCRIPTION
.B duti
sets applications as default handlers for Apple's Uniform Type Identifiers,
for URL schemes, filename extensions, and MIME types.
If
.I settings_path
is not given on the command line,
.B duti
reads settings lines from stdin. If
.I settings_path
is a directory,
.B duti
applies settings from the files in
.IR settings_path .
.sp
The
.BI \-s
flag tells
.B duti
to set a handler based on arguments from the command line. Two arguments
following
.BI \-s
means that
.B duti
will set the handler for a URL scheme. Three arguments means
.B duti
will set the handler for a UTI, an extension or a MIME type, depending on
the formatting of the second argument.
.B duti
treats an argument beginning with a dot as an extension. If the argument
contains no dots,
.B duti
also considers the argument a filename extension, unless it contains a slash,
in which case
.B duti
treats the argument as a MIME type. In all other cases,
.B duti
treats the second argument as a UTI.
.sp
.B duti
.BI \-x
retrieves and prints out information describing the default application
for files with the extension
.IR extension .
.sp
See
.B EXAMPLES
below for usage cases.
.sp
.SH SETTINGS FILE
A settings file is made up of lines with the following format:
.sp
.br
	app_id    UTI    role
.br
.sp
The
.I app_id
is a bundle ID representing the application that will act as the
handler for documents associated with
.IR UTI .
For example:
.sp
.br
	com.apple.Safari	public.html	all
.br
.sp
would cause
.B duti
to set Safari as the default handler in all situations for HTML documents.
A settings file can also contain lines with this format:
.sp
.br
	app_id    url_scheme
.br
.sp
In this case,
.I app_id
is again a bundle ID, this time for the application that will act as the
default handler for
.IR url_scheme .
For example:
.sp
.br
	org.mozilla.Firefox     ftp
.br
.sp
would cause
.B duti
to set Firefox as the handler for "ftp://" URLs.
.SH SETTINGS PLIST
If the extension of the file given to
.B duti
is
.IR \.plist ,
.B duti
treats the file as an XML property list (plist).
The plist must contain a key-value pair, in which the key is "DUTISettings"
and the value is an array of dictionaries. Each dictionary in the array
contains three key-value pairs representing the application's bundle ID,
the UTI and the role, respectively. Alternatively, a dictionary in the array
may contain two key-value pairs representing the application's bundle ID,
and the URL scheme. A simple plist designed to set Safari as the default
handler of HTML files, and Firefox as the default handler for "ftp://" URLs,
would look like this:
.sp
.br
	<?xml version="1.0" encoding="UTF-8"?>
.br
	<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
.br
	<plist version="1.0">
.br
	<dict>
.br
		<key>DUTISettings</key>
.br
		<array>
.br
			<dict>
.br
				<key>DUTIBundleIdentifier</key>
.br
				<string>com.apple.Safari</string>
.br
				<key>DUTIUniformTypeIdentifier</key>
.br
				<string>public.html</string>
.br
				<key>DUTIRole</key>
.br
				<string>all</string>
.br
			</dict>
.br
			<dict>
.br
				<key>DUTIBundleIdentifier</key>
.br
				<string>org.mozilla.Firefox</string>
.br
				<key>DUTIURLScheme</key>
.br
				<string>ftp</string>
.br
			</dict>
.br
		</array>
.br
	</dict>
.br
	</plist>
.br
.sp
.SH ROLES
Valid roles are defined as follows:
.sp
.TP 19
.B all
application handles all roles for the given UTI.
.TP 19
.B viewer
application handles reading and displaying documents with the given UTI.
.TP 19
.B editor
application can manipulate and save the item. Implies viewer.
.TP 19
.B shell
application can execute the item.
.TP 19
.B none
application cannot open the item, but provides an icon for the given UTI.
.SH EXAMPLES
Running
.B duti
with
.BI \-s
:
.sp
.br
	# Set Safari as the default handler for HTML documents
.br
	duti -s com.apple.Safari public.html all
.br
.sp
	# Set Finder as the default handler for the ftp:// URL scheme
.br
	duti -s com.apple.Finder ftp
.sp

Retrieving default application information for an extension:
.sp
.br
	# default application information for .html files
.br
	% duti -x html
.br
	Safari
.br
	/Applications/Safari.app
.br
	com.apple.Safari
.br
.sp

The following examples can be used by passing them to
.B duti
on stdin or as lines in a .duti file.
.sp
Set TextEdit as the default viewer for Microsoft Word documents:
.sp
.br
	com.apple.TextEdit    com.microsoft.word.doc    viewer
.br
.sp
Set VLC as the default viewer for files with .m4v extensions:
.sp
.br
	org.videolan.vlc   m4v	viewer
.br
.sp
Set iHook as the default executor of shell scripts:
.sp
.br
	edu.umich.iHook    public.shell-script    shell
.br
.sp
Set Xcode as the default editor for C source files:
.sp
.br
	com.apple.Xcode    public.c-source    editor
.br
.sp
.SH OPTIONS
.TP 19
.BI \-d\  uti
display the default handler for
.I uti
and exit.
.TP 19
.B \-h
print usage and exit.
.TP 19
.BI \-l\  uti
display all handlers for
.I uti
and exit.
.TP 19
.BI \-s
set the handler from data provided on the command line.
.TP 19
.B \-V
print version number and exit.
.TP 19
.B \-v
verbose output.
.TP 19
.BI \-x\  ext
print information describing the default application for extension
.IR ext .
.sp
.SH EXIT STATUS
.TP 5
0
All settings applied or displayed successfully.
.TP 5
1
Settings could not be applied, or the UTI has no handler.
.TP 5
>1
Error.
.sp
.SH MORE INFO
macOS ships with a number of UTIs already defined. Most third-party
software is responsible for defining its own UTIs. Apple documents UTIs
in the Apple Developer Documentation Archive at:
.sp
.br
	https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/understanding_utis/understand_utis_intro/understand_utis_intro.html
.br
.sp
To get a list of UTIs on your system, you can use the following command line:
.sp
.br
	$(mdfind -name lsregister) -dump | grep -E '^uti:' \\
.br
		| awk '{ print $2 }' | sort | uniq
.sp
.SH SEE ALSO
plutil(1), plist(5)


duti.c
/* duti: set default handlers for document types based on a settings file. */

#include <ApplicationServices/ApplicationServices.h>
#include <CoreFoundation/CoreFoundation.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "handler.h"

extern char		*duti_version;
extern int		nroles;
int			verbose = 0;

struct roles		rtm[] = {
    { "none",   kLSRolesNone },
    { "viewer", kLSRolesViewer },
    { "editor", kLSRolesEditor },
    { "shell",  kLSRolesShell },
    { "all",    kLSRolesAll },
};

    int
main( int ac, char *av[] )
{
    struct stat		st;
    int			c, rc, err = 0;
    int			set = 0;
    int			( *handler_f )( char * );
    char		*path = NULL;
    char		*p;

    extern int		optind;
    extern char		*optarg;

    while (( c = getopt( ac, av, "b:d:e:hl:o:st:u:Vvx:" )) != -1 ) {
	switch ( c ) {
	case 'b':	/* show bundle URLs for identifier */
	    rc = duti_urls_for_bundle( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'd':	/* show default handler for UTI */
	    rc = uti_handler_show( optarg, 0 );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'e':	/* UTI declarations for extension */
	    rc = duti_utis_for_extension( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'h':	/* help */
	default:
	    err++;
	    break;

	case 'l':	/* list all handlers for UTI */
	    rc = uti_handler_show( optarg, 1 );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'o':	/* show all bundles able to open path */
	    rc = duti_urls_for_url( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 's':	/* set handler */
	    set = 1;
	    break;

	case 't':	/* info for type */
	    rc = duti_default_app_for_type( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'u':	/* UTI declarations */
	    rc = duti_utis( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	    break;

	case 'V':	/* version */
	    printf( "%s\n", duti_version );
	    break;

	case 'v':	/* verbose */
	    verbose = 1;
	    break;

	case 'x':	/* info for extension */
	    rc = duti_default_app_for_extension( optarg );
	    if ( rc ) {
	        exit( 0 );
	    }
	}
    }

    if ( !set && !err) {
	exit( 0 );
    }

    nroles = sizeof( rtm ) / sizeof( rtm[ 0 ] );

    switch (( ac - optind )) {
    case 0 :	/* read from stdin */
	if ( set ) {
	    err++;
	}
	break;

    case 1 :	/* read from file or directory */
	if ( set ) {
	    err++;
	}
	path = av[ optind ];
	break;

    case 2 :	/* set URI handler */
	if ( set ) {
	    return( duti_handler_set( av[ optind ],
		    av[ optind + 1 ], NULL ));
	}
	/* this fallthrough works because set == 0 */
	
    case 3 :	/* set UTI handler */
	if ( set ) {
	    return( duti_handler_set( av[ optind ],
		    av[ optind + 1 ], av[ optind + 2 ] ));
	}
	/* fallthrough to error */
    
    default :	/* error */
	err++;
	break;
    }

    if ( err ) {
	fprintf( stderr, "usage: %s [ -hvV ] [ -b bundle_id ] [ -d uti ] [ -e ext ] [ -l uti ] [ -o path ] [ -t type ] [ -u uti ] [ -x ext ]"
		 "[ settings_path ]\n", av[ 0 ] );
	fprintf( stderr, "	-b print the URLs to all bundles with the identifier\n" );
	fprintf( stderr, "	-d print the default handler for the UTI\n" );
	fprintf( stderr, "	-e print the UTI declaration for the file extension\n" );
	fprintf( stderr, "	-l print the handlers for the UTI\n" );
	fprintf( stderr, "	-o print the URLs for all bundles able to open path\n" );
	fprintf( stderr, "	-t print the UTI declaration for the file type\n" );
	fprintf( stderr, "	-u print the UTI declaration for the UTI\n" );
	fprintf( stderr, "	-x print bundle claiming the file extension\n" );
	fprintf( stderr, "usage: %s -s bundle_id url_scheme\n", av[ 0 ] );
	fprintf( stderr, "usage: %s -s bundle_id uti role\n", av[ 0 ] );
	exit( 1 );
    }

    /* by default, read from a FILE stream */
    handler_f = fsethandler;

    if ( path ) {
	if ( stat( path, &st ) != 0 ) {
	    fprintf( stderr, "stat %s: %s\n", path, strerror( errno ));
	    exit( 2 );
	}
	switch ( st.st_mode & S_IFMT ) {
	case S_IFDIR:	/* directory of settings files */
	    handler_f = dirsethandler;
	    break;

	case S_IFREG:	/* settings file or plist */
	    if (( p = strrchr( path, '.' )) != NULL ) {
		p++;
		if ( strcmp( p, "plist" ) == 0 ) {
		    handler_f = psethandler;
		}
	    }
	    break;

	default:
	    fprintf( stderr, "%s: not a supported settings path\n", path );
	    exit( 1 );
	}
    }

    return( handler_f( path ));
}
